docs/architecture/adr-040-storage-and-smt-state-commitments.md

```diff
@@ -138,29 +138,70 @@ for each OP in [Get Has, Set, ...]
+#### Compatibility support
+
+To ease the transition to this new interface for users, we can create a shim which wraps a `CommitMultiStore` but provides a `CommitRootStore` interface, and expose functions to safely create and access the underlying `CommitMultiStore`.
 
-Currently IBC (v1.0) module merkle proof for a `(key, value)` consists of two elements `[storeKey, proof]`. Verification using 2 passes:
-1. Checks that a `proof` is a valid merkle proof `(key, value)` using ICS-23 spec.
-2. Then it checks that the `storeKey` and `root(proof)` hashes to the AppHash (App state commitment).
+The new `RootStore` and supporting types can be implemented in a `store/v2` package to avoid breaking existing code.
+
+#### Merkle Proofs and IBC
 
-Breaking this behavior would severely impact the Cosmos ecosystem which already widely adopts the IBC module. Unfortunately, the straightforward implementation is breaking because all keys in `SC` are hashed. So we can't simply make the last step. Even if we set the `storeKey` to an empty string it will rehash.
+Currently, an IBC (v1.0) Merkle proof path consists of two elements (`["<store-key>", "<record-key>"]`), with each key corresponding to a separate proof. These are each verified according to individual [ICS-23 specs](https://github.com/cosmos/ibc-go/blob/f7051429e1cf833a6f65d51e6c3df1609290a549/modules/core/23-commitment/types/merkle.go#L17), and the result hash of each step is used as the committed value of the next step, until a root commitment hash is obtained.
 
-For workaround we need to:
-+ keep the double hashing and multistore concept for IBC.
-+ the `RootStore` will have only two stores: the general one, and another for IBC. `RootStore` should not expose mounting stores in a "runtime" (it's only possible to do it through constructor).
-+ The App Hash is a hash of both stores in the `RootStore`.
+This is not compatible with the `RootStore`, which stores all records in a single Merkle tree structure, and won't produce separate proofs for the store- and record-key. Ideally, the store-key component of the proof could just be omitted, and updated to use a "no-op" spec, so only the record-key is used. However, because the IBC verification code hardcodes the `"ibc"` prefix and applies it to the SDK proof as a separate element of the proof path, this isn't possible without a breaking change. Breaking this behavior would severely impact the Cosmos ecosystem which already widely adopts the IBC module.
 
+As a workaround, the `RootStore` will have to maintain a separate set of SMT-based stores and an index of store keys, in the same fashion as the `MultiStore`. The non-IBC state can be maintained as a separate store in the index, alongside the modules that are used in IBC (`clients/`, `connections/`, and `channels/`). The final App hash will be the root hash of the store index.
 
+This workaround can be used until the IBC connection code is fully upgraded and supports single-element commitment proofs.
```

Comment:
how about keeping the paragraph: "the `RootStore` will have only two stores....".
Maybe we can be also more specific:
* IBC store key is `01` (2 bits)
* "general" store key is `1` (1 bit).
Not sure though if we can only add as much as 1-2 bits though.

Reply:
Ah, I think I may have misunderstood Aditya's [comment](https://github.com/cosmos/ibc-go/discussions/256#discussioncomment-1013066) I had thought `ibc/clients`, `ibc/connections` were the prefixes of separate stores. Now, if I've got it right, `ibc` is the only store prefix for IBC data, and the rest are just parts of the key paths. I'll update it.

I'm not sure I understand the prefixes you're proposing though. If you mean how the data is partitioned in the DB, that doesn't seem to belong at this level of design, also I don't think it's feasible to have prefixes smaller than a byte - it would shift the alignment of all the following data.

For IBC proofs, I believe the IBC data must be under the `ibc` store key, and the general data can just use the empty string.

Reply:
Yes, my understanding is that there is one store for IBC: the `ibc/`. Then they have a 3 groups of objects: clients, connections, channels.
So my proposal is that in the root store we have 2 stores:
* IBC with `01` prefix (2 bits or 1 byte). AFAIU, IBC supports changing the prefix without doing `x/ibc` update -- it requires a migration only.
* "general" store for everything else.

@AdityaSripal , @colin-axner  - could you validate this?

Reply:
IBC does not support changing the prefix unfortunately because the prefix is in the connection which is non-upgradable.

Can the separate IBC store retain the `ibc/` prefix?

Reply:
yes, if we can't change it then we will keep it.